File: 011 队列的基本实现/顺序表实现.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int front;
    int rear;
    int size;
    int *elems;
} ArrayQueue;

ArrayQueue * createQueue(int size) {
    if (size > 0) {
        ArrayQueue *queue = (ArrayQueue *)malloc(sizeof(ArrayQueue));
        queue->front = 0;
        queue->rear = 0;
        queue->size = size;
        queue->elems = malloc(sizeof(int) * size);
        
        return queue;
    }
    return NULL;
}

int isEmpty(ArrayQueue *queue) {
    return queue->front == queue->rear;
}

int isFull(ArrayQueue *queue) {
    return (queue->rear + 1) % queue->size == queue->front;
}

int length(ArrayQueue *queue) {
    return (queue->rear - queue->front + queue->size) % queue->size;
}

int getHead(ArrayQueue *queue) {
    if (isEmpty(queue)) {
        return -10000;
    }
    return queue->elems[queue->front];
}

void enQueue(ArrayQueue *queue, int value) {
    if (isFull(queue)) {
        return;
    }
    queue->elems[queue->rear] = value;
    queue->rear = (queue->rear + 1) % queue->size;
}

int deQueue(ArrayQueue *queue) {
    if (isEmpty(queue)) {
        return -10000;
    }
    int value = queue->elems[queue->front];
    queue->front = (queue->front + 1) % queue->size;
    return value;
}

void traverse(ArrayQueue *queue) {
    for (int i = queue->front; i != queue->rear; i = (i + 1) % queue->size) {
        printf("%d ", queue->elems[i]);
    }
    printf("\n");
}

void printInfo(ArrayQueue *queue) {
    traverse(queue);
    printf("isEmpty：%d ", isEmpty(queue));
    printf("isFull：%d ", isFull(queue));
    printf("length：%d ", length(queue));
    printf("getHead：%d ", getHead(queue));
    printf("\n");
}

int main() {
    
    ArrayQueue *queue = createQueue(5);
    printInfo(queue);
    
    enQueue(queue, 12);
    enQueue(queue, 34);
    enQueue(queue, 1);
    printInfo(queue);
    
    enQueue(queue, 53);
    enQueue(queue, 109);
    printInfo(queue);
    
    enQueue(queue, 333);
    printInfo(queue);
    
    printf("%d \n", deQueue(queue));
    printInfo(queue);
    
    
    return 0;
}

File: 011 队列的基本实现/顺序表实现.swift
import Foundation

struct ArrayQueue<E> {
    var front = 0
    var rear = 0
    var elems: [E?]
    let size: Int
    
    init(size: Int = 6) {
        self.size = size
        self.elems = Array<E?>(repeating: nil, count: size)
    }
    
    var isEmpty: Bool {
        return front == rear
    }
    
    var isFull: Bool {
        return (rear + 1) % size == front
    }
    
    var length: Int {
        return (rear - front + size) % size
    }
    
    var head: E? {
        if isEmpty {
            return nil
        }
        return elems[front]
    }
    
    mutating func enQueue(_ e: E) {
        if !isFull {
            elems[rear] = e
            rear = (rear + 1) % size
        }
    }
    
    mutating func deQueue() -> E? {
        if !isEmpty {
            let e = elems[front]
            front = (front + 1) % size
            return e
        }
        return nil
    }
    
    func traverse() {
        if !isEmpty {
            var idx = front
            while idx != rear {
                print(elems[idx]!, terminator: " ")
                idx = (idx + 1) % size
            }
            print()
        }
    }
    
    func printInfo() {
        traverse()
        print("isEmpty: \(isEmpty)")
        print("isFull: \(isFull)")
        print("length: \(length)")
        print("head: \(head)")
    }
}

var queue = ArrayQueue<Int>()
queue.printInfo()

queue.enQueue(12)
queue.enQueue(34)
queue.enQueue(3)
queue.printInfo()

queue.enQueue(19)
queue.enQueue(20)
queue.printInfo()

queue.enQueue(440)
queue.printInfo()

print(queue.deQueue())
queue.printInfo()
